File: bot/src/gpt/tool/function/ImageRemoveBackgroundFunction.ts
import AssistantFunction, {AssistantObject, InvocationContext} from "./AssistantFunction";
import {VkMessage} from "../../../service/VkMessagesService";
import {Logger} from "../../../service/LoggingService";
import {Context} from "../../../Context";
import {downloadVkPhotoAttachment} from "./AssistantFunctionHelper";

export default class ImageRemoveBackgroundFunction implements AssistantFunction {
    private logger!: Logger;

    constructor(private context: Context) {
        context.onReady(() => {
            this.logger = context.loggingService.getRootLogger().newSublogger("ImageRemoveBackgroundFunction");
        });
    }

    async call(args: any, message: VkMessage, invocationContext: InvocationContext): Promise<string> {
        this.logger.info("Removing background from image");
        const imageId = args['image_id'];
        await this.context.vkMessagesService.indicateActivity(message.peerId, "photo")
        const image = await downloadVkPhotoAttachment(invocationContext, imageId);
        const resultImage = await this.context.stabilityAiService.removeBackground(image);
        const attachments = await this.context.vkMessagesService.uploadPhotoAttachments(message.peerId, [resultImage]);
        attachments.forEach(attachment => invocationContext.addAttachment(attachment));
        const pseudoMessageText = "[INTERNAL] This is the image with background removed.";
        invocationContext.appendMessage({
            role: "user",
            content: [
                {
                    type: "text",
                    text: pseudoMessageText
                }, {
                    type: "image_url",
                    image_url: {
                        url: "data:image/png;base64," + resultImage.toString('base64')
                    }
                }
            ]
        });
        invocationContext.chargeCredits(3);
        return "Image is attached to message.";
    }

    getDescription(): string {
        return "Removes background from an image";
    }

    getName(): string {
        return "image_remove_background";
    }

    getParameters(): AssistantObject | null {
        return {
            type: "object",
            properties: {
                image_id: {
                    type: "string",
                    description: "ID of the image to remove background from"
                }
            },
            required: ["image_id"]
        }
    }

}

File: backend/src/main/java/space/davids_digital/sweetie/gpt/tool/function/ImageRemoveBackgroundFunction.kt
package space.davids_digital.sweetie.gpt.tool.function

import com.aallam.openai.api.chat.ChatMessage
import com.aallam.openai.api.chat.ImagePart
import com.aallam.openai.api.chat.TextPart
import com.vk.api.sdk.objects.messages.SetActivityType
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import space.davids_digital.sweetie.gpt.InvocationContext
import space.davids_digital.sweetie.gpt.tool.function.AssistantFunctionUtils.downloadPhotoAttachment
import space.davids_digital.sweetie.gpt.tool.function.parameter.Description
import space.davids_digital.sweetie.integration.openai.OpenAiUtils.toBase64PngDataUrl
import space.davids_digital.sweetie.integration.stabilityai.StabilityAiService
import space.davids_digital.sweetie.integration.vk.VkMessageService
import space.davids_digital.sweetie.model.VkMessageModel

@Component
class ImageRemoveBackgroundFunction(
    private val vkMessagesService: VkMessageService,
    private val stabilityAiService: StabilityAiService
): AssistantFunction<ImageRemoveBackgroundFunction.ImageRemoveBackgroundParameters> {
    companion object {
        private val log = LoggerFactory.getLogger(ImageRemoveBackgroundFunction::class.java)
    }

    data class ImageRemoveBackgroundParameters(
        @Description("ID of the image to remove background from")
        val imageId: Int
    )

    override fun getName() = "image_remove_background"
    override fun getDescription() = "Removes background from an image using Stable Diffusion"
    override fun getParametersClass() = ImageRemoveBackgroundParameters::class

    override suspend fun call(
        parameters: ImageRemoveBackgroundParameters,
        message: VkMessageModel,
        invocationContext: InvocationContext,
    ): String {
        log.info("Removing background from image with id ${parameters.imageId}")
        vkMessagesService.indicateActivity(message.peerId, SetActivityType.PHOTO)
        val inputImage = invocationContext.downloadPhotoAttachment(parameters.imageId)
        val resultImage = stabilityAiService.removeBackground(inputImage)
        val attachments = vkMessagesService.uploadPhotoAttachment(message.peerId, resultImage)
        invocationContext.addAttachment(attachments)
        invocationContext.appendMessage(
            ChatMessage.Companion.User(listOf(
                TextPart("[INTERNAL] This is the result image:"),
                ImagePart(resultImage.toBase64PngDataUrl())
            )))
        invocationContext.chargeCredits(3)
        return "Image is attached to message."
    }
}